File: bindings/javascript/src/jvm.js
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

import {
  DatalevinConfigurationError,
  DatalevinJavaError,
  DatalevinJvmError
} from "./errors.js";

export const DATALEVIN_JAR_ENV = "DATALEVIN_JAR";
export const DATALEVIN_CLASSPATH_ENV = "DATALEVIN_CLASSPATH";
export const DATALEVIN_JVM_ARGS_ENV = "DATALEVIN_JVM_ARGS";
export const DATALEVIN_JAVACPP_CACHEDIR_ENV = "DATALEVIN_JAVACPP_CACHEDIR";
export const DEFAULT_JAVACPP_CACHEDIR = "/tmp/datalevin-javacpp-cache";
export const DEFAULT_JVM_ARGS = [
  "--enable-native-access=ALL-UNNAMED",
  "--add-opens=java.base/java.lang=ALL-UNNAMED",
  "--add-opens=java.base/java.util=ALL-UNNAMED",
  "--add-opens=java.base/java.nio=ALL-UNNAMED",
  "--add-opens=java.base/sun.nio.ch=ALL-UNNAMED"
];

const TARGET_JAR_PATTERNS = [
  /^datalevin-runtime-.*\.jar$/,
  /^datalevin-python-runtime-.*\.jar$/,
  /^datalevin-java-.*\.jar$/
];

const packageRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const jarsDir = path.join(packageRoot, "jars");
const repoRoot = path.resolve(packageRoot, "../..");
const targetDir = path.join(repoRoot, "target");

let bridgeModulePromise = null;
let startJvmPromise = null;
let classesPromise = null;
let started = false;

function expandPath(value) {
  if (value.startsWith("~")) {
    return path.resolve(os.homedir(), value.slice(1));
  }
  return path.resolve(value);
}

function findRuntimeJars(dir) {
  if (!fs.existsSync(dir)) {
    return [];
  }

  return fs.readdirSync(dir)
    .filter((entry) => TARGET_JAR_PATTERNS.some((pattern) => pattern.test(entry)))
    .sort()
    .map((entry) => path.join(dir, entry));
}

function splitShellWords(input) {
  if (!input) {
    return [];
  }

  const words = [];
  let current = "";
  let quote = null;
  let escape = false;

  for (const ch of input) {
    if (escape) {
      current += ch;
      escape = false;
      continue;
    }

    if (ch === "\\") {
      escape = true;
      continue;
    }

    if (quote !== null) {
      if (ch === quote) {
        quote = null;
      } else {
        current += ch;
      }
      continue;
    }

    if (ch === "'" || ch === "\"") {
      quote = ch;
      continue;
    }

    if (/\s/.test(ch)) {
      if (current) {
        words.push(current);
        current = "";
      }
      continue;
    }

    current += ch;
  }

  if (escape || quote !== null) {
    throw new DatalevinConfigurationError(
      `Invalid ${DATALEVIN_JVM_ARGS_ENV} shell-style argument string.`
    );
  }

  if (current) {
    words.push(current);
  }

  return words;
}

function ensureJavacppCachedirArg(jvmArgs) {
  for (const arg of jvmArgs) {
    if (
      arg.startsWith("-Dorg.bytedeco.javacpp.cachedir=") ||
      arg.startsWith("-Dorg.bytedeco.javacpp.cacheDir=")
    ) {
      return jvmArgs;
    }
  }

  const cacheDir = process.env[DATALEVIN_JAVACPP_CACHEDIR_ENV] || DEFAULT_JAVACPP_CACHEDIR;
  fs.mkdirSync(cacheDir, { recursive: true });
  jvmArgs.push(`-Dorg.bytedeco.javacpp.cachedir=${cacheDir}`);
  return jvmArgs;
}

function ensureDefaultJvmArgs(jvmArgs) {
  for (const arg of DEFAULT_JVM_ARGS) {
    if (!jvmArgs.includes(arg)) {
      jvmArgs.push(arg);
    }
  }
  return jvmArgs;
}

function resolveJvmArgs(jvmArgs = null) {
  let resolved;
  if (jvmArgs && jvmArgs.length > 0) {
    resolved = [...jvmArgs];
  } else {
    const envArgs = process.env[DATALEVIN_JVM_ARGS_ENV];
    resolved = splitShellWords(envArgs);
  }

  return ensureJavacppCachedirArg(ensureDefaultJvmArgs(resolved));
}

export function resolveClasspath(classpath = null) {
  if (classpath && classpath.length > 0) {
    return classpath.map((entry) => expandPath(entry));
  }

  const envClasspath = process.env[DATALEVIN_CLASSPATH_ENV];
  if (envClasspath) {
    return envClasspath
      .split(path.delimiter)
      .filter(Boolean)
      .map((entry) => expandPath(entry));
  }

  const envJar = process.env[DATALEVIN_JAR_ENV];
  if (envJar) {
    return [expandPath(envJar)];
  }

  const vendored = findRuntimeJars(jarsDir);
  if (vendored.length > 0) {
    return vendored;
  }

  const repoLocal = findRuntimeJars(targetDir);
  if (repoLocal.length > 0) {
    return repoLocal;
  }

  throw new DatalevinConfigurationError(
    "Unable to resolve a Datalevin runtime jar. "
      + `Set ${DATALEVIN_JAR_ENV}, set ${DATALEVIN_CLASSPATH_ENV}, `
      + "or vendor a jar under bindings/javascript/jars/."
  );
}

export function jvmStarted() {
  return started;
}

export async function javaBridgeModule() {
  if (bridgeModulePromise !== null) {
    return bridgeModulePromise;
  }

  bridgeModulePromise = import("java-bridge")
    .then((module) => module.default ?? module)
    .catch((error) => {
      bridgeModulePromise = null;
      throw new DatalevinConfigurationError(
        "The `java-bridge` package is required for the Datalevin Node bindings.",
        { cause: error }
      );
    });
  return bridgeModulePromise;
}

function javaClassName(value) {
  try {
    return value.getClassSync().getNameSync();
  } catch {
    return null;
  }
}

function javaMessage(value) {
  try {
    return value.getMessageSync();
  } catch {
    return null;
  }
}

function coerceJavaError(error) {
  const cause = error?.cause && typeof error.cause.getClassSync === "function" ? error.cause : null;
  const javaClass = cause === null ? null : javaClassName(cause);
  const typeName =
    cause !== null && typeof cause.getErrorTypeSync === "function"
      ? cause.getErrorTypeSync()
      : null;
  const data =
    cause !== null && typeof cause.getDataSync === "function"
      ? cause.getDataSync()
      : null;

  return new DatalevinJavaError(
    javaMessage(cause) || error?.message || "Java call failed.",
    {
      javaClass,
      typeName,
      data,
      cause: error
    }
  );
}

export async function callJava(target, ...args) {
  try {
    return await target(...args);
  } catch (error) {
    throw coerceJavaError(error);
  }
}

export async function callJavaMethod(receiver, methodName, ...args) {
  return callJava(receiver[methodName].bind(receiver), ...args);
}

export async function startJvm({ classpath = null, jvmArgs = null } = {}) {
  if (startJvmPromise !== null) {
    return startJvmPromise;
  }

  startJvmPromise = (async () => {
    const bridge = await javaBridgeModule();
    const resolvedClasspath = resolveClasspath(classpath);
    const resolvedJvmArgs = resolveJvmArgs(jvmArgs);

    try {
      bridge.ensureJvm({
        classpath: resolvedClasspath,
        opts: resolvedJvmArgs
      });
      started = true;
    } catch (error) {
      startJvmPromise = null;
      throw new DatalevinJvmError(
        "Failed to start the JVM for Datalevin. "
          + "Check that Java 21+ is installed and that the runtime jar is valid.",
        { cause: error }
      );
    }
  })();

  return startJvmPromise;
}

export async function classes() {
  if (classesPromise !== null) {
    return classesPromise;
  }

  classesPromise = (async () => {
    await startJvm();
    const { importClass } = await javaBridgeModule();
    const opts = { asyncJavaExceptionObjects: true };
    const load = (className) => importClass(className, opts);

    return {
      datalevin: load("datalevin.Datalevin"),
      interop: load("datalevin.DatalevinInterop"),
      udfFunction: load("datalevin.UdfFunction"),
      datalevinException: load("datalevin.DatalevinException"),
      linkedHashMap: load("java.util.LinkedHashMap"),
      linkedHashSet: load("java.util.LinkedHashSet"),
      arrayList: load("java.util.ArrayList"),
      longClass: load("java.lang.Long"),
      uuid: load("java.util.UUID"),
      instant: load("java.time.Instant"),
      date: load("java.util.Date"),
      bigInteger: load("java.math.BigInteger"),
      bigDecimal: load("java.math.BigDecimal"),
      mapType: load("java.util.Map"),
      listType: load("java.util.List"),
      setType: load("java.util.Set"),
      collectionType: load("java.util.Collection"),
      keywordType: load("clojure.lang.Keyword"),
      symbolType: load("clojure.lang.Symbol")
    };
  })().catch((error) => {
    classesPromise = null;
    throw error;
  });

  return classesPromise;
}

export const __testing = {
  findRuntimeJars,
  splitShellWords,
  resolveJvmArgs,
  ensureDefaultJvmArgs,
  ensureJavacppCachedirArg
};

File: bindings/python/src/datalevin/_jvm.py
"""JVM bootstrap and runtime jar resolution."""

from __future__ import annotations

import importlib.resources as resources
import os
import re
import shlex
from pathlib import Path

import jpype

from .errors import DatalevinConfigurationError, DatalevinJvmError

DATALEVIN_JAR_ENV = "DATALEVIN_JAR"
DATALEVIN_CLASSPATH_ENV = "DATALEVIN_CLASSPATH"
DATALEVIN_JVM_ARGS_ENV = "DATALEVIN_JVM_ARGS"
DATALEVIN_JAVACPP_CACHEDIR_ENV = "DATALEVIN_JAVACPP_CACHEDIR"
PACKAGE_NAME = "datalevin"
TARGET_JAR_PATTERNS = (
    "datalevin-runtime-*.jar",
    "datalevin-python-runtime-*.jar",
    "datalevin-java-*.jar",
)
DEFAULT_JAVACPP_CACHEDIR = Path("/tmp/datalevin-javacpp-cache")
DEFAULT_JVM_ARGS = (
    "--enable-native-access=ALL-UNNAMED",
    "--add-opens=java.base/java.lang=ALL-UNNAMED",
    "--add-opens=java.base/java.util=ALL-UNNAMED",
    "--add-opens=java.base/java.nio=ALL-UNNAMED",
    "--add-opens=java.base/sun.nio.ch=ALL-UNNAMED",
)


def jvm_started() -> bool:
    """Return whether the current Python process already has a JVM."""

    return jpype.isJVMStarted()


def resolve_classpath(classpath: list[str] | tuple[str, ...] | None = None) -> list[str]:
    """Resolve the JVM classpath for the Datalevin runtime."""

    if classpath:
        return [str(Path(entry).expanduser()) for entry in classpath]

    env_classpath = os.environ.get(DATALEVIN_CLASSPATH_ENV)
    if env_classpath:
        return [entry for entry in env_classpath.split(os.pathsep) if entry]

    env_jar = os.environ.get(DATALEVIN_JAR_ENV)
    if env_jar:
        return [str(Path(env_jar).expanduser())]

    vendored = _vendored_jars()
    if vendored:
        return vendored

    repo_local = _repo_local_jars()
    if repo_local:
        return repo_local

    raise DatalevinConfigurationError(
        "Unable to resolve a Datalevin runtime jar. "
        f"Set {DATALEVIN_JAR_ENV}, set {DATALEVIN_CLASSPATH_ENV}, "
        "or vendor a jar under datalevin/jars/."
    )


def start_jvm(
    *,
    classpath: list[str] | tuple[str, ...] | None = None,
    jvm_args: list[str] | tuple[str, ...] | None = None,
    convert_strings: bool = False,
) -> None:
    """Start the JVM if it is not already running."""

    if jpype.isJVMStarted():
        return

    resolved_classpath = resolve_classpath(classpath)
    resolved_jvm_args = list(jvm_args or ())
    if not resolved_jvm_args:
        env_args = os.environ.get(DATALEVIN_JVM_ARGS_ENV)
        if env_args:
            resolved_jvm_args = shlex.split(env_args)
    _ensure_default_jvm_args(resolved_jvm_args)
    _ensure_javacpp_cachedir_arg(resolved_jvm_args)

    try:
        jpype.startJVM(
            *resolved_jvm_args,
            classpath=resolved_classpath,
            convertStrings=convert_strings,
        )
    except OSError as exc:
        raise DatalevinJvmError(
            "Failed to start the JVM for Datalevin. "
            "Check that Java 21+ is installed and that the runtime jar is valid.",
            cause=exc,
        ) from exc


def _vendored_jars() -> list[str]:
    try:
        jar_root = resources.files(PACKAGE_NAME).joinpath("jars")
    except (FileNotFoundError, ModuleNotFoundError):
        return []

    if not jar_root.is_dir():
        return []

    jar = _preferred_runtime_jar(Path(str(jar_root)))
    return [] if jar is None else [str(jar)]


def _repo_local_jars() -> list[str]:
    repo_root = Path(__file__).resolve().parents[4]
    target_dir = repo_root / "target"
    if not target_dir.exists():
        return []
    jar = _preferred_runtime_jar(target_dir)
    return [] if jar is None else [str(jar)]


def _preferred_runtime_jar(dir_path: Path) -> Path | None:
    for pattern in TARGET_JAR_PATTERNS:
        matches = sorted(dir_path.glob(pattern), key=_jar_sort_key)
        if matches:
            return matches[-1]
    return None


def _jar_sort_key(path: Path):
    stem = path.stem
    version = stem.rsplit("-", 1)[-1]
    parts = re.split(r"([0-9]+)", version)
    return [int(part) if part.isdigit() else part for part in parts]


def _ensure_javacpp_cachedir_arg(jvm_args: list[str]) -> None:
    for arg in jvm_args:
        if (
            arg.startswith("-Dorg.bytedeco.javacpp.cachedir=")
            or arg.startswith("-Dorg.bytedeco.javacpp.cacheDir=")
        ):
            return

    cache_dir = Path(os.environ.get(DATALEVIN_JAVACPP_CACHEDIR_ENV, DEFAULT_JAVACPP_CACHEDIR))
    cache_dir.mkdir(parents=True, exist_ok=True)
    jvm_args.append(f"-Dorg.bytedeco.javacpp.cachedir={cache_dir}")


def _ensure_default_jvm_args(jvm_args: list[str]) -> None:
    for arg in DEFAULT_JVM_ARGS:
        if arg not in jvm_args:
            jvm_args.append(arg)
